uniffi_bindgen/src/bindings/kotlin/templates/ObjectTemplate.kt

```diff
@@ -267,20 +283,52 @@ open class {{ impl_class_name }}: Disposable, AutoCloseable, {{ interface_name }
 {% include "CallbackInterfaceImpl.kt" %}
 {%- endif %}
 
+{%- let trait_impl=format!("uniffiCallbackInterface{}", name) %}
+
+{%- if obj.has_callback_interface() %}
 /**
  * @suppress
  */
-public object {{ ffi_converter_name }}: FfiConverter<{{ type_name }}, Pointer> {
-    {%- if obj.has_callback_interface() %}
+public object {{ ffi_converter_name }}: FfiConverterRustBuffer<{{ type_name }}> {
     internal val handleMap = UniffiHandleMap<{{ type_name }}>()
-    {%- endif %}
+    internal val langIndex: Int = {{ trait_impl }}.register(UniffiLib.INSTANCE)
+
+    override fun read(buf: ByteBuffer): {{ type_name }} {
+        // The Rust code always writes pointers as 8 bytes, and will
+        // fail to compile if they don't fit.
+        val handle = buf.getLong()
+        val langIndex = buf.getInt()
+        if (handleMap.has(handle)) {
+            return handleMap.get(handle)
+        } else {
+            return {{ impl_class_name }}(Pointer(handle))
+        }
```

Comment:
I'm not sure that this right yet: we should be able to detect here if the trait was implemented in our language, Rust or another foreign language.